kotest-framework/kotest-framework-multiplatform-plugin-gradle/build.gradle.kts

```diff
@@ -23,47 +25,28 @@ repositories {
 }
 
 dependencies {
-   compileOnly(gradleApi())
-   compileOnly(libs.kotlin.gradle.plugin)
+   implementation(libs.kotlin.gradle.plugin)
 
+   testImplementation(gradleTestKit())
    testImplementation(project(Projects.Assertions.Core))
    testImplementation(project(Projects.Framework.api))
    testImplementation(project(Projects.Framework.engine))
    testImplementation(project(Projects.JunitRunner))
 }
 
-tasks.withType<Test> {
-   // Build these libraries ahead of time so that the test project doesn't try to build them itself (if it tries to build them while we are as well, this can lead to conflicts)
-   setOf(
-      Projects.Assertions.Core,
-      Projects.Framework.api,
-      Projects.Framework.engine
-   ).forEach { project ->
-      setOf(
-         "jvmJar",
-         "compileKotlinLinuxX64",
-         "compileKotlinMacosX64",
-         "compileKotlinMacosArm64",
-         "compileKotlinMingwX64",
-      ).forEach { task ->
-         dependsOn("$project:$task")
-      }
-   }
-
-   setOf(
-      Projects.JunitRunner,
-      ":kotest-framework:kotest-framework-multiplatform-plugin-embeddable-compiler",
-      ":kotest-framework:kotest-framework-multiplatform-plugin-legacy-native"
-   ).forEach { project ->
-      dependsOn("$project:jvmJar")
-   }
-
-   dependsOn("jar")
-   dependsOn(":kotlinNpmInstall")
-
+tasks.withType<Test>().configureEach {
    useJUnitPlatform()
 
    systemProperty("kotestVersion", Ci.publishVersion)
+   val gradleWrapper = if ("windows" in System.getProperty("os.name").toLowerCase()) {
```

Comment:
I would strongly recommend against invoking the Gradle wrapper from the tests, and instead encourage retaining the task dependencies. 

If these tasks are run from within the tests, Gradle may end up running the same tasks at the same time (once from the original Gradle invocation that is running the tests at the same time, and again from the Gradle invocation from inside the tests). This can lead to conflicts, weird errors and all other kinds of pain.

Reply:
I agree that manually invoking the Gradle wrapper twice is not ideal (TestKit would be better), but I don't see how the PR is different to the current code, could you explain? The Gradle wrapper script is still invoked from within the test the same way. But It's better now though because the test code doesn't need to know about the project structure to discover the Gradle wrapper, and it works on all OSes.

Reply:
Ah, I think I attached my comment to the wrong part of the diff sorry. Passing through the path to the Gradle wrapper is fine - my comment was about the removal of all of the task dependencies for the test tasks and then running `./gradlew build -x check` in the tests.

Reply:
Well, depending on the output of other projects by referencing them isn't best practice either :)

> Don’t reference other project tasks directly
> https://docs.gradle.org/current/userguide/cross_project_publications.html

I've worked on a way to make the Gradle TestKit work well, I think this would be the best way:  https://github.com/kotest/kotest/pull/3163

Reply:
> Well, depending on the output of other projects by referencing them isn't best practice either :)

Totally :) This was the best way I could find to ensure that Gradle wasn't building Kotest in two processes at once (once in the Gradle instance running the plugin tests, and another time as a result of the included build into the dummy Gradle project used by the tests). If there's a better and more reliable way of doing this, then go for it.

> I've worked on a way to make the Gradle TestKit work well, I think this would be the best way: https://github.com/kotest/kotest/pull/3163

Awesome - using TestKit wasn't an option when I added the test originally due to the issue described in #3059 (Gradle embedded an old version of Kotlin, which caused problems running the tests).

Can this PR (#3156) be merged independently of #3163? Is there some way to break these changes up to make them smaller and easier to review? Both of these PRs are quite large.

Reply:
I've broken down this PR into smaller ones - see the main comment. 

> Awesome - using TestKit wasn't an option when I added the test originally due to the issue described in https://github.com/kotest/kotest/issues/3059 (Gradle embedded an old version of Kotlin, which caused problems running the tests).

Yeah, the embedded Kotlin version is a pain to work around. It's frustrating that the Gradle team is so slow to update it. Although [7.5 does use Kotlin 1.6.21,](https://docs.gradle.org/current/userguide/compatibility.html#kotlin) so that should match Kotest....

Hmm actually I suspect what was wrong is that currently on master branch the compiler plugin sets the Kotlin Kotlin language level to 1.6. Gradle basically requires the language level is 1.4. I added the `kotlin-dsl` plugin, which sets the language level to 1.4, thus resolving the issues.